File: Language/C#/Book/tmp/LINQ实战/Chapter06to08.CSharp/Common/SampleHarness/SampleClass.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Text;

namespace LinqInAction.Chapter06to08.Common.SampleHarness
{
    public class SampleItem
    {
        public int Chapter { get; set; }
        public int ListingNumber { get; set; }
        public string Description { get; set; }
        public MethodInfo Method { get; set; }
    }

    public class SampleClass
        : IEnumerable<SampleItem>
    {
        List<SampleItem> _MethodList;
        String _Title;
        StreamWriter _OutputStreamWriter;

        public SampleClass()
        {
            _OutputStreamWriter = new StreamWriter(new MemoryStream());

            Type ClassType = base.GetType();
            this._Title = ClassType.Name;

            var items = from MethodInfo _method in ClassType.GetMethods(BindingFlags.Public | BindingFlags.Instance | BindingFlags.DeclaredOnly)
                        from SampleAttribute _attrib in _method.GetCustomAttributes(false).OfType<SampleAttribute>()
                        select new SampleItem
                        {
                            Method = _method,
                            Description = _attrib.Description,
                            Chapter = _attrib.Chapter,
                            ListingNumber = _attrib.ListingNumber
                        };

            _MethodList = new List<SampleItem>();
            _MethodList.AddRange(items);
        }

        #region IEnumerable<SampleItem> Members

        public IEnumerator<SampleItem> GetEnumerator()
        {
            return _MethodList.GetEnumerator();
        }

        #endregion

        //#region IEnumerable Members

        //IEnumerator IEnumerable.GetEnumerator()
        //{
        //    return _MethodList.GetEnumerator();
        //}

        //#endregion

        public SampleItem this[int index]
        {
            get
            {
                return this._MethodList[index];
            }
        }

        public string Title
        {
            get { return _Title; }
        }

        public StreamWriter OutputStreamWriter
        {
            get { return _OutputStreamWriter; }
        }

        #region IEnumerable Members

        IEnumerator IEnumerable.GetEnumerator()
        {
            return _MethodList.GetEnumerator();
        }

        #endregion
    }
}

File: Language/C#/Book/tmp/LINQ实战/Chapter06to08.VB/Common/SampleHarness/SampleClass.vb
Imports System.Reflection
Imports System.IO

Public Class SampleItem
    Private _Chapter As Integer
    Private _ListingNumber As Integer
    Private _Description As String
    Private _Method As MethodInfo

    'Public Sub New(ByVal method As MethodInfo)
    '    _Method = method
    '    _Description = method.Name
    'End Sub

    Public Property Chapter() As Integer
        Get
            Return _Chapter
        End Get
        Set(ByVal value As Integer)
            _Chapter = value
        End Set
    End Property

    Public Property ListingNumber() As Integer
        Get
            Return _ListingNumber
        End Get
        Set(ByVal value As Integer)
            _ListingNumber = value
        End Set
    End Property

    Public Property Description() As String
        Get
            Return _Description
        End Get
        Set(ByVal value As String)
            _Description = value
        End Set
    End Property

    Public Property Method() As MethodInfo
        Get
            Return _Method
        End Get
        Set(ByVal value As MethodInfo)
            _Method = value
        End Set
    End Property

End Class

Public Class SampleClass
    Implements IEnumerable(Of SampleItem), IEnumerable

    Private _MethodList As New List(Of SampleItem)
    Private _Title As String

    Public Sub New()
        Dim ClassType As Type = MyBase.GetType
        _Title = ClassType.Name

        Dim items = From _method As MethodInfo In ClassType.GetMethods(BindingFlags.Public Or BindingFlags.Instance Or BindingFlags.DeclaredOnly), _
                    _attrib As SampleAttribute In _method.GetCustomAttributes(False).OfType(Of SampleAttribute)() _
                   Select New SampleItem With { _
                                               .Method = _method, _
                                               .Description = _attrib.Description, _
                                               .Chapter = _attrib.Chapter, _
                                               .ListingNumber = _attrib.ListingNumber _
                                               }

        _MethodList.AddRange(items)

    End Sub

    Public Function GetEnumerator1() As System.Collections.Generic.IEnumerator(Of SampleItem) Implements System.Collections.Generic.IEnumerable(Of SampleItem).GetEnumerator
        Return _MethodList.GetEnumerator
    End Function

    Public Function GetEnumerator() As System.Collections.IEnumerator Implements System.Collections.IEnumerable.GetEnumerator
        Return _MethodList.GetEnumerator
    End Function

    Default Public ReadOnly Property item(ByVal index As Integer) As SampleItem
        Get
            Return _MethodList(index)
        End Get
    End Property

    Public ReadOnly Property Title() As String
        Get
            Return _Title
        End Get
    End Property

    Public ReadOnly Property OutputStreamWriter() As StreamWriter
        Get
            Return _OutputStreamWriter
        End Get
    End Property
    Private ReadOnly _OutputStreamWriter As StreamWriter = New StreamWriter(New MemoryStream)

End Class